tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,248 @@
+#                        MemoryFormat? memory_format=None) -> Tensor
+# - Argument Conversion
+#       Generates C++ code to convert an ivalue (from stack) to its underlying C++ type.
+#    - int[] size
+#        ```cpp
+#           const c10::List<c10::IValue> size_list_in = (std::move(peek(stack, 0, 7))).toList();
+#
+#           std::vector<int64_t> size_vec;
+#           for (c10::IValue size_elem: size_list_in) {
+#               int64_t size_base = size_elem.to<int64_t>();
+#               size_vec.push_back(size_base);
+#           }
+#           at::ArrayRef<int64_t> size_list_out(size_vec);
+#                                 ~~~~~~~~~~~~~ <-- The converted argument from ivalues in the stack.
+#                                                   Will be passed to unboxed kernel.
+#       ```
+#    - Dimname[]? names
+#       ```cpp
+#           c10::optional<c10::IValue> names_opt = (std::move(peek(stack, 1, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ArrayRef<at::Dimname>> names_opt_out;
+#           if (names_opt.has_value()) {
+#                         ~~~~~~~~~~~ <-- Unwrapping optional shell
+#               const c10::IValue names_opt_in = names_opt.value();
+#               const c10::List<c10::IValue> names_list_in = names_opt_in.toList();
+#
+#               std::vector<at::Dimname> names_vec;
+#               for (c10::IValue names_elem: names_list_in) {
+#                                ~~~~~~~~~~~~~~~~~~~~~~~~~ <-- Unrolling list, then convert elements one by one.
+#                   at::Dimname names_base = names_elem.to<at::Dimname>();
+#                   names_vec.push_back(names_base);
+#               }
+#               at::ArrayRef<at::Dimname> names_list_out(names_vec);
+#
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>(names_list_out);
+#           } else {
+#               names_opt_out = c10::optional<at::ArrayRef<at::Dimname>>();
+#           }
+#       ```
+#    - ScalarType? dtype (similarly for the rest of the arguments)
+#       ```cpp
+#           c10::optional<c10::IValue> dtype_opt = (std::move(peek(stack, 2, 7))).toOptional<c10::IValue>();
+#           c10::optional<at::ScalarType> dtype_opt_out;
+#           if (dtype_opt.has_value()) {
+#               const c10::IValue dtype_opt_in = dtype_opt.value();
+#               at::ScalarType dtype_base = dtype_opt_in.to<at::ScalarType>();
+#                                                        ~~~~~~~~~~~~~~~~~~~~ <-- For base types, convert ivalue to it
+#                                                                                 directly using ".to<T>()" API.
+#               dtype_opt_out = c10::optional<at::ScalarType>(dtype_base);
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#    - (implicitly) TensorOptions
+#       ```cpp
+#           const auto options = TensorOptions()
```

Comment:
Do you need the `TensorOptions` for anything?

Some background: the C++ API (the codegen'd ops that look like `at::{op}`) come in two flavors, the "nonfaithful" and "faithful" versions. For the non-faithful API:
- factory functions get "TensorOptions" in their signature
- default arguments show up as defaulted in the signature
- out= operators receive the output tensor as their first argument

For the "faithful" API:
- TensorOptions args are exploded into their individual pieces
- default arguments in the yaml are not defaulted
- out= ops receive their output tensor args at the end.

The "non-faithful" API exists partially as a convenience C++ API: it's more convenient in C++ to user TensorOptions for example, and have defaults. the "faithful" API is better for performance, since it aligns directly with other internal API's (for example, the actual C++ kernels don't know about TensorOptions; they have to be unwrapped).

In this case, I don't think you need TensorOptions, because iiuc, boxed schemas already have the exploded, individual factory args in their schema, and not TensorOptions. That means that the unboxing will be cleaner to write, and also faster at runtime, if you don't involve TensorOptions.

Reply:
Are you saying the TensorOptions will be unwrapped at the codegen time and the elements in the stack will always be `layout`, `device` etc?

Reply:
It seems like that's already the case, looking at your codegen output? The logic you have here is popping each argument off the stack individual (`layout`, `device`, etc), and then wrapping them all into a `TensorOptions` argument.

So it seems like the individual arguments are already on the stack. And then if that's true, then there's no need to re-wrap then into TensorOptions when calling into the C++ API.

Reply:
Got it. Let me try to remove the code generating `TensorOptions` and see if it still compiles.